### PR TITLE
[5.1] Use ::class constant

### PR DESCRIPTION
**Pull request by JosephSilber**
Makes it far less error-prone and infinitely more readable - especially in the tests.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
I already said don't do this because it adds little value to the framework. It only makes sense for laravel/laravel.

File: src/Illuminate/Auth/AuthServiceProvider.php [hunk 1]
```diff
@@ -48,7 +49,7 @@ protected function registerAuthenticator()
 	 */
 	protected function registerUserResolver()
 	{
-		$this->app->bind('Illuminate\Contracts\Auth\Authenticatable', function($app)
+		$this->app->bind(Authenticatable::class, function($app)
```
Comment (GrahamCampbell) [inline src/Illuminate/Auth/AuthServiceProvider.php]: Does this even work for interfaces?

**Comment by JosephSilber** [comment]
> I already said don't do this

No idea where/when you said that, but...

I totally disagree that it "adds little value to the framework". As I said, it makes code far less error-prone and infinitely more readable - of value everywhere.

File: src/Illuminate/Auth/AuthServiceProvider.php (hunk 1 above)
Comment (JosephSilber) [inline src/Illuminate/Auth/AuthServiceProvider.php]: It sure does.

**Comment by GrahamCampbell** [comment]
https://github.com/laravel/framework/issues/8954

**Comment by GrahamCampbell** [comment]
> As I said, it makes code far less error-prone 

Not if you've made any mistakes in this PR it doesn't. It's hard for us to check because you've changed the tests too.

**Comment by GrahamCampbell** [comment]
Also, a single commit is fine.

File: src/Illuminate/Auth/AuthServiceProvider.php (hunk 1 above)
Comment (GrahamCampbell) [inline src/Illuminate/Auth/AuthServiceProvider.php]: Great. Just looks strange to have "class" for an interface, lol.

**Comment by JosephSilber** [comment]
> It's hard for us to check because you've changed the tests too

The changes are in separate commits. Very easy to check.

File: src/Illuminate/Auth/AuthServiceProvider.php (hunk 1 above)
Comment (crynobone) [inline src/Illuminate/Auth/AuthServiceProvider.php]: Does `$app->make('Illuminate\Contracts\Auth\Authenticatable') === $app->make('\Illuminate\Contracts\Auth\Authenticatable')`? Saw some issue while back with some of this inconsistency (but not sure if it was ever fixed).

**Comment by Anahkiasen** [comment]
I can confirm this approach is less error prone, at least refactor wise. Far easier to update use statements that to look in every single string. Even PHPStorm trips a lot on strings whereas he can mass update use statements throughout a whole codebase.

File: src/Illuminate/Auth/AuthServiceProvider.php (hunk 1 above)
Comment (GrahamCampbell) [inline src/Illuminate/Auth/AuthServiceProvider.php]: It should do. `'\Illuminate\Contracts\Auth\Authenticatable'` technically isn't the name of that class though. The fully qualified name is without the slash.

**Comment by GrahamCampbell** [comment]
> at least refactor wise

Yeh, but we're not doing any refactoring.

**Comment by Anahkiasen** [comment]
I don't think you can honestly garantuee no class will ever be renamed or moved, it's not like it hasn't happened before.

**Comment by Anahkiasen** [comment]
Even disregarding refactors you're pushing for classes dependencies to be listed at the top of the file instead of using their FQN (which I approve of), to me this is in the continuity of that.

**Comment by GrahamCampbell** [comment]
@Anahkiasen I'm not really against this to be honest, but equally I'm not particularly for it either. We shouldn't change things for the sake of it as large changes like this do carry risk of error.

**Comment by franzliedke** [comment]
That's what we have large releases such as 5.1 for, right?

**Comment by taylorotwell** [comment]
I've said many times we don't accept these types of massive PRs unless they are far planned ahead of time (such as 5.1 moving to PSR2). It's just not practical to read through every file and every line to try to catch subtle little bugs. We can update to ::class slowly as we make changes to the framework or updates to a particular test file.

**Pull request closed without merging**